### PR TITLE
[TypeTraits] Move general-purpose metaprogramming utilities from TDataFrame's utils to ROOT TypeTraits

### PR DESCRIPTION
**Pull request by bluehood**
Here it is...any kind of feedback welcome :)

### CODE REVIEW

**Comment by bluehood** [comment]
A google test is on the way

File: core/foundation/inc/ROOT/TypeTraits.hxx [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+/*************************************************************************
+ * Copyright (C) 1995-2017, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef ROOT_TTypeTraits
+#define ROOT_TTypeTraits
```
Comment (amadio) [inline core/foundation/inc/ROOT/TypeTraits.hxx]: Isn't this supposed to drop the extra T as well?

Comment (bluehood) [inline core/foundation/inc/ROOT/TypeTraits.hxx]: good catch, thanks!

**Comment by bluehood** [comment]
A google test has been added in `$ROOTSYS/core/foundation/test`. `roottest_dataframe_functiontraits` has been broken by moving the meta utils away from TDataFrame, and is removed by PR n.48 in roottest.

**Comment by dpiparo** [comment]
@phsft-bot build please.

File: core/foundation/inc/ROOT/TypeTraits.hxx [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+};
+
+/// Check whether a histogram type is a classic or v7 histogram.
+template <typename T>
+struct IsV7Hist : public std::false_type {
+   static_assert(std::is_base_of<TH1, T>::value, "not implemented for this type");
+};
+
+template <int D, typename P, template <int, typename, template <typename> class> class... S>
+struct IsV7Hist<ROOT::Experimental::THist<D, P, S...>> : public std::true_type {
```
Comment (pcanal) [inline core/foundation/inc/ROOT/TypeTraits.hxx]: Could we move these two back to an internal file, it is not quite 'general' enough :) [ i.e. ROOT implementation detail]

File: core/foundation/inc/ROOT/TypeTraits.hxx [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+
+template <typename... Args>
+using RemoveFirst_t = typename RemoveFirst<Args...>::type;
+
+/// Return first of possibly many template parameters.
+/// For non-template types, the result is the type itself.
+/// e.g. TakeFirstParameter<U<A,B>> is A
+///      TakeFirstParameter<T> is T
+template <typename T>
+struct TakeFirstParameter {
```
Comment (pcanal) [inline core/foundation/inc/ROOT/TypeTraits.hxx]: What is the difference between TakeFirst and TakeFirstParameter?  Do we need both?  Can we distinguish them more?

File: core/foundation/inc/ROOT/TypeTraits.hxx (hunk 2 above)
Comment (bluehood) [inline core/foundation/inc/ROOT/TypeTraits.hxx]: we don't want ROOT type traits in ROOT/TypeTraits? :) note that it does not introduce a dependency on libHisto, it just needs a forward decl

File: core/foundation/test/TypeTraits.cxx [hunk 4]
```diff
@@ -0,0 +1,143 @@
+#include "ROOT/TypeTraits.hxx"
```
Comment (pcanal) [inline core/foundation/test/TypeTraits.cxx]: Can we please rename the test file either TypeTraits.test.cxx (or testTypeTrait.cxx) to distinguish it (beyhond just the directory) from an implementation file?  Thanks.

Comment (bluehood) [inline core/foundation/test/TypeTraits.cxx]: sure

File: core/foundation/inc/ROOT/TypeTraits.hxx (hunk 3 above)
Comment (bluehood) [inline core/foundation/inc/ROOT/TypeTraits.hxx]: `TakeFirst` acts on variadic lists of types, e.g. `TakeFirst_t<A,B>` is `A`. `TakeFirstParameter` acts on the parameter list of a template object, e.g. `TakeFirstParameter_t<U<A,B>>` is `A`. They do different things, so I would say we "need" them both. I'm open to suggestions regarding the names (`TakeFirstType` and `TakeFirstParameter`?). I thought the docs were clear but maybe not?

File: core/foundation/inc/ROOT/TypeTraits.hxx [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+struct TypeList {
+   static constexpr std::size_t list_size = sizeof...(Types);
+};
+
+/// Extract types from the signature of a callable object.
+template <typename T>
+struct CallableTraits {
+   using arg_types = typename CallableTraits<decltype(&T::operator())>::arg_types;
+   using arg_types_nodecay = typename CallableTraits<decltype(&T::operator())>::arg_types_nodecay;
+   using ret_type = typename CallableTraits<decltype(&T::operator())>::ret_type;
```
Comment (Axel-Naumann) [inline core/foundation/inc/ROOT/TypeTraits.hxx]: Let me rephrase:
```
struct foo{
  int operator()(int);
  double operator()(double);
};
using res = typename CallableTraits<foo(float)>::ret_type;
```
should `res` be `double` or should this be ill-formed?

File: core/foundation/inc/ROOT/TypeTraits.hxx [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+   using type = TypeList<Rest...>;
+};
+
+template <typename... Args>
+using RemoveFirst_t = typename RemoveFirst<Args...>::type;
+
+/// Return first of possibly many template parameters.
+/// For non-template types, the result is the type itself.
+/// e.g. TakeFirstParameter<U<A,B>> is A
+///      TakeFirstParameter<T> is T
```
Comment (Axel-Naumann) [inline core/foundation/inc/ROOT/TypeTraits.hxx]: That is a very ... rare ... behavior. Providing this as a general tool might see misuses / misunderstanding of the interface. I'd naively expect that `TakeFirstParameter<T>` is either empty or ill-formed, instead of mixing template levels... Do you expect generic usage of this? (Do you really know you need this behavior in your use case? I didn't look at it, it seems surprising...)

File: core/foundation/inc/ROOT/TypeTraits.hxx [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+struct RemoveFirstParameter<U<T, Rest...>> {
+   using type = U<Rest...>;
+};
+
+template <typename T>
+using RemoveFirstParameter_t = typename RemoveFirstParameter<T>::type;
+
+/// Compile-time integer sequence generator
+/// e.g. calling GenStaticSeq<3>::type() instantiates a StaticSeq<0,1,2>
+// TODO deprecate as soon as std::index_sequence is backported
```
Comment (Axel-Naumann) [inline core/foundation/inc/ROOT/TypeTraits.hxx]: Can we move this into `Experimental::`? Then we can just remove it without deprecation, and for now, no other interfaces should be using it.

File: core/foundation/inc/ROOT/TypeTraits.hxx [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+class IsSmartOrDumbPtr<std::unique_ptr<P>> : public std::true_type {
+};
+
+/// Check for container traits.
+template <typename T>
+struct IsContainer {
+   using Test_t = typename std::decay<T>::type;
+
+   template <typename A>
+   static constexpr bool Test(A *pt, A const *cpt = nullptr, decltype(pt->begin()) * = nullptr,
```
Comment (bluehood) [inline core/foundation/inc/ROOT/TypeTraits.hxx]: Thanks for looking into this. This was implemented by @dpiparo so I would ask his opinion too.

File: core/foundation/inc/ROOT/TypeTraits.hxx [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+   {
+      return false;
+   }
+
+   static constexpr bool value = Test<Test_t>(nullptr);
+};
+
+/// Check whether a histogram type is a classic or v7 histogram.
+template <typename T>
+struct IsV7Hist : public std::false_type {
```
Comment (bluehood) [inline core/foundation/inc/ROOT/TypeTraits.hxx]: @pcanal expressed the same concern. I'm outnumbered, so I'll move `IsV7Hist` back inside `TDataFrame`'s stuff.

File: core/foundation/inc/ROOT/TypeTraits.hxx (hunk 5 above)
Comment (bluehood) [inline core/foundation/inc/ROOT/TypeTraits.hxx]: What is the standard wording? `result_of` and `invoke_result` contain a `type`, but I think it would be a confusing name here? boost uses `return_type` in their `function_traits`, I can change it to that if you prefer.

Your example should be ill-formed because `CallableTraits` does not require to specify the argument types, users just pass in a callable object. Supporting the syntax you used might be a nice addition but we have no use for it, at least for now.

File: core/foundation/inc/ROOT/TypeTraits.hxx (hunk 6 above)
Comment (bluehood) [inline core/foundation/inc/ROOT/TypeTraits.hxx]: I agree, I will change the behaviour

File: core/foundation/test/TypeTraits.cxx (hunk 4 above)
Comment (vgvassilev) [inline core/foundation/test/TypeTraits.cxx]: We were using EntityToTest`Tests.cxx` pattern. Having something/test/test*.cxx looks very weird...

File: core/foundation/inc/ROOT/TypeTraits.hxx (hunk 7 above)
Comment (bluehood) [inline core/foundation/inc/ROOT/TypeTraits.hxx]: I am never sure when it comes to ROOT namespaces. Do you mean `ROOT::Experimental::` or `ROOT::TypeTraits::Experimental::` or `ROOT::Experimental::TypeTraits::`?

File: core/foundation/inc/ROOT/TypeTraits.hxx [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+template <typename R, typename T, typename... Args>
+struct CallableTraits<R (T::*)(Args...) const> {
+   using arg_types = TypeList<typename std::decay<Args>::type...>;
+   using arg_types_nodecay = TypeList<Args...>;
+   using ret_type = R;
+};
+
+// mutable lambdas and functor classes
+template <typename R, typename T, typename... Args>
+struct CallableTraits<R (T::*)(Args...)> {
```
Comment (bluehood) [inline core/foundation/inc/ROOT/TypeTraits.hxx]: Can you give me an example?

This works:
```
struct A { void foo() {} };
static_assert(std::is_same<CallableTraits<decltype(&A::foo)>::ret_type, void>::value, "");
```

Comment (Axel-Naumann) [inline core/foundation/inc/ROOT/TypeTraits.hxx]: Does `static_assert(std::is_same<CallableTraits<decltype(A::foo)>::ret_type, void>::value, "");` work?

File: core/foundation/inc/ROOT/TypeTraits.hxx (hunk 5 above)
Comment (Axel-Naumann) [inline core/foundation/inc/ROOT/TypeTraits.hxx]: Forget what I said - you test the callable, not the expression. Return type is just fine.

File: core/foundation/test/TypeTraits.cxx (hunk 4 above)
Comment (Axel-Naumann) [inline core/foundation/test/TypeTraits.cxx]: Who's "we"? `ls */*/test/*.cxx` seems to disagree...

File: core/foundation/inc/ROOT/TypeTraits.hxx (hunk 10 above)
Comment (bluehood) [inline core/foundation/inc/ROOT/TypeTraits.hxx]: It does not. Can you ever use `decltype(A::foo)` anywhere?

File: core/foundation/inc/ROOT/TypeTraits.hxx (hunk 6 above)
Comment (bluehood) [inline core/foundation/inc/ROOT/TypeTraits.hxx]: done

File: core/foundation/inc/ROOT/TypeTraits.hxx (hunk 9 above)
Comment (bluehood) [inline core/foundation/inc/ROOT/TypeTraits.hxx]: done

File: core/foundation/inc/ROOT/TypeTraits.hxx (hunk 3 above)
Comment (bluehood) [inline core/foundation/inc/ROOT/TypeTraits.hxx]: renamed to `TakeFirstType`

File: core/foundation/inc/ROOT/TypeTraits.hxx (hunk 2 above)
Comment (bluehood) [inline core/foundation/inc/ROOT/TypeTraits.hxx]: moved back to TDataFrame

File: core/foundation/inc/ROOT/TypeTraits.hxx (hunk 10 above)
Comment (Axel-Naumann) [inline core/foundation/inc/ROOT/TypeTraits.hxx]: Indeed, thanks! Forget this comment, too! :-)

File: core/foundation/inc/ROOT/TypeTraits.hxx [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+
+/// Lightweight storage for a collection of types.
+/// Differently from std::tuple, no instantiation of objects of stored types is performed
+template <typename... Types>
+struct TypeList {
+   static constexpr std::size_t list_size = sizeof...(Types);
+};
+
+/// Extract types from the signature of a callable object.
+template <typename T>
```
Comment (bluehood) [inline core/foundation/inc/ROOT/TypeTraits.hxx]: will do

File: core/foundation/inc/ROOT/TypeTraits.hxx (hunk 3 above)
Comment (pcanal) [inline core/foundation/inc/ROOT/TypeTraits.hxx]: I suppose your renamed the first one (formely TakeFirst), in which case, this is indeed better.  thanks

File: core/foundation/inc/ROOT/TypeTraits.hxx (hunk 7 above)
Comment (pcanal) [inline core/foundation/inc/ROOT/TypeTraits.hxx]: Actually since it is not meant to be used by users (std::index_sequence is), I would put it in ::ROOT::Internal (and move it back to DataFrame unless there is another 'current' user ...

Comment (bluehood) [inline core/foundation/inc/ROOT/TypeTraits.hxx]: I'll move it back and push for having a backport of std::index_sequence soon :)

**Comment by bluehood** [comment]
Addressed most of the comments, cleaned up history (for some reason github does not show it as it is on file).

TODO:
- make `CallableTraits` SFINAE-friendly (thanks Axel)
- remove `type` from the most generic `TakeFirstParameter`, have `TDataFrame` use something else

**Comment by bluehood** [comment]
@Axel-Naumann sorry I can't seem to produce a reasonable use-case for the sfinae-friendliness of `CallableTraits`, could you please provide a little example snippet of how you would like it to work?

**Comment by Axel-Naumann** [comment]
@bluehood > reasonable use-case for the sfinae-friendliness of `CallableTraits`

What about `TMVA::AddDataSource("name", T)`. That `T` is either an array of `ELEM`, or something returning `ELEM`. Would be nice if I can use `CallableTraits` to figure out the return type - if it's a function. Instead of having to hide this in another template that is then sfinae protecting the "o but your array has no call operator!" error.

As I said during our chat: I don't know whether that's a sufficiently strong reason to make `CallableTraits` sfinae friendly. But as you're publishing this to the world now you could as well make it super nice ;-)

**Comment by bluehood** [comment]
@Axel-Naumann I am not able to write a sfinae-friendly version of `CallableTraits`. As far as I understand, your proposed usage would be the following:

```c++
template<typename T, std::size_t N>           
void foo(std::array<T,N> a) { return; }
                                                                                                                        
template<typename F, typename R = typename CallableTraits<F>::ret_type>  
void foo(F f) { return; }                       
                                                                                                                        
int main() {                                                                                                                       
   foo(std::array<int,1>{1}); // overload #2 is SFINAE'd out
   foo([]() { return std::array<int,1>{1}; }); // overload #1 does not apply                                       
   return 0;                                             
}    
```

The self-contained snippet [attached](https://github.com/root-project/root/files/1082588/testcallabletraits.txt) was my playground to try and make this work, with no success.

In principle I would like to have an empty `CallableTraits` as the most generic object, and then partially specialise for the types `T` that have `T::operator()`, but afaik it's not possible to partially specialise on properties of a type. The second idea was writing two mutually exclusive (SFINAE-wise) `template<typename T> CallableTraits`, an empty one for all types without `operator()`, and the usual one for types with `operator()`, but I don't know how to write a SFINAE expression that answers the question "does T have operator()?" with a boolean.

It's very possible that this would be trivial with concepts. Any suggestion is super welcome.

**Comment by bluehood** [comment]
P.S. there is a [google groups conversation](https://groups.google.com/forum/#!msg/comp.lang.c++.moderated/T3x6lvmvvkQ/mfY5VTDJ--UJ) that points to an article in russian where something like this has apparently been implemented....

**Comment by Axel-Naumann** [comment]
https://wandbox.org/permlink/vEWSUerAwZpQCdIm

**Comment by bluehood** [comment]
well damn :)

**Comment by bluehood** [comment]
The last two items on my to-do list have been addressed. Branch has been rebased on master to avoid failing on newly added tests.

**Comment by bluehood** [comment]
Rebased and fixed clang-formatting. There are no items pending afaik.

File: core/foundation/inc/ROOT/TypeTraits.hxx (hunk 8 above)
Comment (dpiparo) [inline core/foundation/inc/ROOT/TypeTraits.hxx]: Hi @bluehood , @Axel-Naumann I am of course fine with the new implementation!

**Comment by bluehood** [comment]
Rebased on master, resolved conflicts
